File: elements-of-programming-interviews/2023/python/string_integer_interconversion.py
from test_framework import generic_test
from test_framework.test_failure import TestFailure


def int_to_string(x: int) -> str:
    if not x:
        return "0"
    neg = "-" if x < 0 else ""
    x = x*-1 if neg else x
    string = []
    while x:
        string.append(str(x % 10))
        x //= 10
    return neg + "".join(string[::-1])


def string_to_int(s: str) -> int:
    place = 1
    val = 0
    neg = s[0] == "-"
    s = s[1:] if neg or s[0] == "+" else s
    for char in s[::-1]:
        val += int(char) * place
        place *= 10
    return val * -1 if neg else val


def wrapper(x, s):
    if int(int_to_string(x)) != x:
        raise TestFailure('Int to string conversion failed')
    if string_to_int(s) != x:
        raise TestFailure('String to int conversion failed')


if __name__ == '__main__':
    # print(int_to_string(0))
    # print(string_to_int("0"))
    # """
    exit(
        generic_test.generic_test_main('string_integer_interconversion.py',
                                       'string_integer_interconversion.tsv',
                                       wrapper))
    # """

File: elements-of-programming-interviews/2023/cpp/string_integer_interconversion.cc
#include <string>

#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
using std::string;

#define ASCII_ZERO 48

string IntToString(int x) {
  if (!x) {
    return "0";
  }
  bool neg = (x < 0);
  string val;
  while (x) {
    int digit = abs(x % 10) + ASCII_ZERO;
    val += digit;
    x /= 10;
  }
  if (neg) {
    val += "-";
  }
  reverse(val.begin(), val.end());
  return val;
}

int StringToInt(const string& s) {
  bool neg = (s[0] == '-');
  bool skipFirstChar = neg || (s[0] == '+');
  unsigned place = 1;
  int total = 0;
  for (int i = s.length() - 1; i >= (skipFirstChar ? 1 : 0); i--) {
    char digit = s[i] - ASCII_ZERO;
    total += (digit * place);
    place *= 10;
  }
  return total * (neg ? -1 : 1);
}

void Wrapper(int x, const string& s) {
  if (stoi(IntToString(x)) != x) {
    throw TestFailure("Int to string conversion failed");
  }

  if (StringToInt(s) != x) {
    throw TestFailure("String to int conversion failed");
  }
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"x", "s"};
  return GenericTestMain(args, "string_integer_interconversion.cc",
                         "string_integer_interconversion.tsv", &Wrapper,
                         DefaultComparator{}, param_names);
}
